### ISSUE TITLE
Relationships among model classes

### ISSUE THREAD

**Issue by Sergey-Nosov-GH** [polterguy/magic#7]
In real projects, of all but simplest complexity, model classes often relate to each other via one-to-one, one-to-many, or many-to many navigational properties.

Perhaps it would be a good idea to add a master-detail example to the project. Or at least some guidance, on where the best extensibility points are to implement models with navigational properties.

Thank you.

**Comment by polterguy** [comment]
That's a problem I have thought extensively about, without really having the time, or a struck of inspiration to fix - But yes, that is a very important fact you are pinpointing here.

My own use case example for this, in my own head, is that for instance the _"magic.email"_ module, would probably want to create a link between its `EmailAccount` domain type to the `User` domain type from _"magic.auth"_ - This would of course break the pure modular design of these modules, creating references from _"magic.email"_ to _"magic.auth"_, eliminating its _"perfect modularity"_.

However, in a real world example, creating such an association, to ensure referential integrity between an email account and a user, would be an obvious example of where creating such a reference across modules would be justified to perfectly model your domain model.

In fact, such problems as you describe here, is why I have not chosen to create NuGet packages out of the thing, and arguably also why Magic seems to violate the _"open/closed"_ principle of SOLID programming, allowing its code to be modified, instead of consumed as a _"closed/extendible module"_.

However, after a while, I realised that I was not building _"libraries"_ per se (which are perfect candidates for NuGet packages), but rather _"starter kits"_, intended to provide a sound initial _"structure"_ to start out new projects, with the intention of having its users modify the code (violating the _"closed"_ principle).

Some other parts of the library are still candidates for NuGet packages, such as for instance the HttpClient and the Signal implementation, since they would never require modification - But there I kind of concluded with that by providing the code, instead of a NuGet package, I could (hopefully) teach junior developers something about the importance of _"clean code"_, and such _"inspire"_ them to avoid creating _"wet code"_ violating coding standards, sound architectural principles, etc ...

Besides the HTtpClient might also require modification, to allow for consuming non-JSON types of payloads, etc ...

However, yes! In a real world application, some of the modularity in Magic would have to be thrown out, to allow for modelling your domain types the way you illustrate. I am not sure how I could do something that would in a general way illustrate this, and/or allow for this, without _"destroying the thing's value"_.

Maybe what you're looking for is maybe just some additional documentation about this?

Something illustrating _"real world usage"_ where I actually create such references ...?

**Comment by polterguy** [comment]
FYI, the Signals implementation could in some cases reduce the numbers of these cross module references, since it allows for invoking methods in other assemblies, without creating references between any of the parties.

In an early version of the code, I raised a _"super signal"_ when a user was deleted, which I would handle in service implementation of the _"magic.email"_ module. This allowed for me to delete all email accounts belonging to a specific user, without having a reference between the email and the auth module at all.

However, I realised it made the code too complex, and the correct way to achieve this, would still be to create a one-to-many reference from user to email accounts.

But for some scenarios, this is a perfectly valid solution, avoiding cross module references, by instead simply creating a signal handler and raise a signal when some event occurs in another module.
